lib/parser/openmp-grammar.h

```diff
@@ -0,0 +1,336 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPENMP_PARSER_GRAMMAR_H_
+#define OPENMP_PARSER_GRAMMAR_H_
+
+// Top-level grammar specification for Fortran.  These parsers drive
+// the tokenization parsers in cooked-tokens.h to consume characters,
+// recognize the productions of Fortran, and to construct a parse tree.
+// See ParserCombinators.md for documentation on the parser combinator
+// library used here to implement an LL recursive descent recognizer.
+
+#include "basic-parsers.h"
+#include "characters.h"
+#include "debug-parser.h"
+#include "grammar.h"
+#include "parse-tree.h"
+#include "stmt-parser.h"
+#include "token-parsers.h"
+#include "type-parsers.h"
+#include "user-state.h"
+#include <cinttypes>
+#include <cstdio>
+#include <functional>
+#include <list>
+#include <optional>
+#include <string>
+#include <tuple>
+#include <utility>
+
+// OpenMP Directives and Clauses
+namespace Fortran::parser {
+
+constexpr auto beginOmpDirective = skipEmptyLines >> space >> "!$OMP "_sptok;
+
+// OpenMP Clauses
+
+// DEFAULT (PRIVATE | FIRSTPRIVATE | SHARED | NONE )
+TYPE_PARSER(construct<OmpDefaultClause>(
+    "PRIVATE" >> pure(OmpDefaultClause::Type::Private) ||
+    "FIRSTPRIVATE" >> pure(OmpDefaultClause::Type::Firstprivate) ||
+    "SHARED" >> pure(OmpDefaultClause::Type::Shared) ||
+    "NONE" >> pure(OmpDefaultClause::Type::None)))
+
+// PROC_BIND(CLOSE | MASTER | SPREAD)
+TYPE_PARSER(construct<OmpProcBindClause>(
+    "CLOSE" >> pure(OmpProcBindClause::Type::Close) ||
+    "MASTER" >> pure(OmpProcBindClause::Type::Master) ||
+    "SPREAD" >> pure(OmpProcBindClause::Type::Spread)))
+
+// MAP ([ [map-type-modifier[,]] map-type : ] list)
+// map-type-modifier -> ALWAYS
+// map-type -> TO | FROM | TOFROM | ALLOC | RELEASE | DELETE
+TYPE_PARSER(construct<OmpMapClause>(
+    maybe(maybe("ALWAYS"_tok >> maybe(","_ch)) >>
+        ("TO" >> pure(OmpMapClause::Type::To) / ":"_ch ||
```

Comment:
Why are some strings that represent tokens user-defined string literals with `_tok` and others not?

Reply:
I will fix them to be all token, space is optional.

Reply:
`"xyz"_tok` can be written as `"xyz"` before '>>' or after `/`.

Be advised, `":"_ch` doesn't skip spaces before or after the character.  `"xyz"_ch` will match a single character from the given set without skipping spaces.  Maybe some of these single character matches should be token matches.

Reply:
Thanks Peter for the explanation.  Yes, I will will fix "xx"_ch too.

Reply:
Is the  documentation on the parser combinator library unclear or too difficult to find?  We should make sure that programmers outside PGI will have an easy time understanding this code, so please let me know if you think that the documentation could be improved.

Reply:
It is my oversight on "xx"_ch.  Thank you.